File: 2020/day21/day21.js
const fs = require("fs");
const runner = require("../../utils/runner");

const IN = fs.readFileSync("21.in", "utf8");

const ingredients = {};
function pt1() {
  const pt1 = [];
  IN.split("\n").forEach((line) => {
    let [items, allergens] = line.split(" (contains ");
    items = items.split(" ");
    pt1.push(items);
    allergens = allergens.slice(0, -1).split(", ");
    allergens.forEach((allgn) => {
      const fltrd = items.filter((itm) =>
        ingredients[allgn] ? ingredients[allgn].has(itm) : true
      );
      ingredients[allgn] = new Set(fltrd);
    });
  });

  const done = new Set();
  while (Object.values(ingredients).some((s) => s.size > 1)) {
    for (const [k, v] of Object.entries(ingredients)) {
      if (v.size === 1 && !done.has(k)) {
        const [ingrdnt] = v.values();
        for (const [K, V] of Object.entries(ingredients)) {
          if (V.size > 1 && K !== k) {
            V.delete(ingrdnt);
          }
        }
        done.add(k);
      }
    }
  }

  const foundAllergens = Object.values(ingredients).reduce(
    (acc, curr) => acc.concat(Array.from(curr)),
    []
  );

  const ans = pt1.reduce(
    (acc, curr) => acc + curr.filter((i) => !foundAllergens.includes(i)).length,
    0
  );
  return ans;
}

function pt2() {
  const sorted = Object.keys(ingredients)
    .sort()
    .map((item) => Array.from(ingredients[item])[0]);
  return sorted.join(",");
}

runner(1, pt1);
runner(2, pt2);

File: 2020/day21/day21.py
import re, fileinput
from collections import Counter

D = dict()
C = Counter()
for line in fileinput.input():
    ingredients, allergens = re.findall("(.*?)(?: \\((?:contains )(.+?)\\))", line)[0]
    ingredients = set(ingredients.split(" "))
    C += Counter(ingredients)
    for allergen in allergens.split(", "):
        if allergen in D:
            D[allergen] &= ingredients
        else:
            D[allergen] = ingredients.copy()
            
bad = set().union(*D.values())
pt1 = sum(v for k, v in C.items() if k not in bad)

pt2 = []
while D:
    k, v = [(k, v) for k, v in D.items() if len(v) == 1][0]
    del D[k]
    for K, V in D.items():
        D[K] = V - v
    pt2.append((k, v.pop()))
pt2.sort(key=lambda x: x[0])

print(pt1, ",".join(x[1] for x in pt2))
